src/mmc.c

```diff
@@ -0,0 +1,272 @@
+
+#include "plugin.h"
+#include "utils/common/common.h"
+#include "utils/ignorelist/ignorelist.h"
+
+#if !KERNEL_LINUX
+#error "No applicable input method."
+#endif
+
+#define PLUGIN_NAME "mmc"
+#define SYS_PATH "/sys/bus/mmc/devices/"
+
+#define DEVICE_KEY "Device"
+#define IGNORE_KEY "IgnoreSelected"
+
+static const char *config_keys[] = {
+    DEVICE_KEY,
+    IGNORE_KEY,
+};
+static int config_keys_num = STATIC_ARRAY_SIZE(config_keys);
+
+#define MMC_MANUFACTOR "manfid"
+#define MMC_OEM_ID "oemid"
+#define MMC_SSR "ssr"
+
+static ignorelist_t *ignorelist = NULL;
+
+static int mmc_config(const char *key, const char *value) {
+  if (ignorelist == NULL)
+    ignorelist = ignorelist_create(1);
+
+  if (ignorelist == NULL) {
+    ERROR(PLUGIN_NAME ": Ignorelist_create failed");
+    return -ENOMEM;
+  }
+
+  if (strcasecmp(key, DEVICE_KEY) == 0) {
+    if (ignorelist_add(ignorelist, value)) {
+      ERROR(PLUGIN_NAME ": Cannot add value to ignorelist");
+      return -1;
+    }
+  } else if (strcasecmp(key, IGNORE_KEY) == 0)
+    ignorelist_set_invert(ignorelist, IS_TRUE(value) ? 0 : 1);
+  else {
+    ERROR(PLUGIN_NAME ": Invalid option %s", key);
+    return -1;
+  }
+
+  return 0;
+}
+
+static void mmc_submit(const char *dev_name, const char *type, gauge_t value) {
+  value_list_t vl = VALUE_LIST_INIT;
+
+  vl.values = &(value_t){.gauge = value};
+  vl.values_len = 1;
+  sstrncpy(vl.plugin, PLUGIN_NAME, sizeof(vl.plugin));
+  sstrncpy(vl.type_instance, dev_name, sizeof(vl.type_instance));
+  sstrncpy(vl.type, type, sizeof(vl.type));
+
+  plugin_dispatch_values(&vl);
+}
+
+static int mmc_read_dev_attr(const char *dev_name, const char *file_name,
+                             char *buffer, int size) {
+  FILE *fh;
+  char str[sizeof(SYS_PATH) + strlen(dev_name) + sizeof("/") +
+           strlen(file_name) + 1];
+  int length;
+
+  snprintf(str, sizeof(str), SYS_PATH "%s/%s", dev_name, file_name);
+  fh = fopen(str, "r");
+
+  if (fh == NULL) {
+    ERROR(PLUGIN_NAME "(%s): Cannot open file [%s]", dev_name, str);
+    return EXIT_FAILURE;
+  }
+
+  DEBUG(PLUGIN_NAME "(%s): try to read [%s]", dev_name, str);
+  if (fgets(buffer, size, fh) == NULL) {
+    ERROR(PLUGIN_NAME "(%s): Unable to read file [%s] (%s)", dev_name, str,
+          STRERRNO);
+    fclose(fh);
+    return EXIT_FAILURE;
+  }
+  fclose(fh);
+
+  /* Remove trailing whitespace for sysfs attr read */
+  length = strlen(buffer);
+  DEBUG(PLUGIN_NAME "(%s): Read %d characters [%s]", dev_name, length, str);
+  if (buffer > 0)
+    buffer[length - 1] = '\0';
+
+  return 0;
+}
+
+static int mmc_read_manfid(const char *dev_name, int *value) {
+  char buffer[4096];
+
+  if (mmc_read_dev_attr(dev_name, MMC_MANUFACTOR, buffer, sizeof(buffer)) ==
+      0) {
+    *value = (int)strtol(buffer, NULL, 0);
+    DEBUG(PLUGIN_NAME "(%s): [%s]=%s (%d)", dev_name, MMC_MANUFACTOR, buffer,
+          *value);
+    return 0;
+  }
+
+  WARNING(PLUGIN_NAME "(%s): Unable to read manfid", dev_name);
```

Comment:
Can you please make this warning a bit more speaking? Not everybody knows what a "manfid" is...